File: src/test/java/pl/patrykdolata/lifemanager/mapper/UserMapperTest.java
package pl.patrykdolata.lifemanager.mapper;

import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.security.crypto.password.PasswordEncoder;
import pl.patrykdolata.lifemanager.domain.UserEntity;
import pl.patrykdolata.lifemanager.model.NewUser;

import java.util.Date;

import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.*;
import static org.junit.jupiter.api.Assertions.*;

@SpringBootTest
public class UserMapperTest {

    @Autowired
    private UserMapper mapper;

    @Autowired
    private PasswordEncoder passwordEncoder;

    @Test
    public void mapPassword_whenEmptyPassword_expectRandomString() {
        String password = "";
        String result = mapper.mapPassword(password);

        assertPasswordLength(result);
    }

    @Test
    public void mapPassword_whenNullPassword_expectRandomString() {
        String password = null;
        String result = mapper.mapPassword(password);

        assertPasswordLength(result);
    }

    @Test
    public void mapPassword_whenSpecificPassword_expectProperString() {
        String password = "test_password";
        String result = mapper.mapPassword(password);

        assertPasswordLength(result);
        assertTrue(passwordEncoder.matches(password, result));
    }

    private void assertPasswordLength(String result) {
        assertNotNull(result);
        assertThat(result, hasLength(60));
    }

    @Test
    public void toEntity_whenNullModel_expectNull() {
        NewUser user = null;
        UserEntity result = mapper.toEntity(user);

        assertNull(result);
    }

    @Test
    public void toEntity_whenProperModel_expectProperEntity() {
        NewUser user = new NewUser("username", "pass", "pass", "test@test.com",
                "John", "Doe");
        UserEntity result = mapper.toEntity(user);

        assertAll(
                () -> assertNotNull(result),
                () -> assertThat(result.getUsername(), is("username")),
                () -> assertThat(result.getFirstName(), is("John")),
                () -> assertThat(result.getLastName(), is("Doe")),
                () -> assertThat(result.getEmail(), is("test@test.com")),
                () -> assertFalse(result.getActivated()),
                () -> assertThat(result.getActivationKey(), hasLength(UserMapper.KEY_LENGTH)),
                () -> assertThat(result.getCreatedAt().doubleValue(), closeTo(new Date().getTime(), 60000.0))
        );
    }
}

File: src/test/kotlin/pl/patrykdolata/lifemanager/mapper/UserMapperTest.kt
package pl.patrykdolata.lifemanager.mapper

import org.hamcrest.MatcherAssert.assertThat
import org.hamcrest.Matchers.*
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertThrows
import org.junit.jupiter.api.function.Executable
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.security.crypto.password.PasswordEncoder
import pl.patrykdolata.lifemanager.model.NewUser
import java.util.*

@SpringBootTest
class UserMapperTest {

    @Autowired
    private lateinit var mapper: UserMapper

    @Autowired
    private lateinit var passwordEncoder: PasswordEncoder

    @Test
    fun `Mapping empty password should return random encrypt password`() {
        val password = ""
        val result = mapper.mapPassword(password)
        assertPasswordLength(result)
    }

    @Test
    fun `Mapping null password should throw NPE`() {
        val password: String? = null
        assertThrows<NullPointerException> { mapper.mapPassword(password!!) }
    }

    @Test
    fun `Mapping proper password should return specific encrypt password string`() {
        val password = "test_password"
        val result = mapper.mapPassword(password)
        assertPasswordLength(result)
        assertTrue(passwordEncoder.matches(password, result))
    }

    private fun assertPasswordLength(result: String) {
        assertNotNull(result)
        assertThat(result, hasLength(60))
    }

    @Test
    fun `Mapping null NewUser should throw NPE`() {
        val user: NewUser? = null
        assertThrows<NullPointerException> { mapper.toEntity(user!!) }
    }

    @Test
    fun `Mapping proper NewUser should return proper UserEntity`() {
        val user = NewUser("username", "pass", "pass", "test@test.com",
                "John", "Doe")
        val result = mapper.toEntity(user)
        assertAll(
                Executable { assertNotNull(result) },
                Executable { assertThat(result.username, equalTo("username")) },
                Executable { assertThat(result.firstName, equalTo("John")) },
                Executable { assertThat(result.lastName, equalTo("Doe")) },
                Executable { assertThat(result.email, equalTo("test@test.com")) },
                Executable { assertTrue(result.activated) },
                Executable { assertThat(result.activationKey, hasLength(UserMapper.KEY_LENGTH)) },
                Executable { assertThat(result.createdAt.toDouble(), closeTo(Date().time.toDouble(), 60000.0)) }
        )
    }
}
